File: Doubao_MarsCode/go/MC0049/main.go
package main

import (
	"fmt"
	"sort"
)

func solution(n int, s []string, x []int) []string {
	merged := make(map[string]int)
	orderMap := make(map[string]int)

	for i := 0; i < n; i++ {
		name := s[i]
		amount := x[i]

		merged[name] += amount

		if _, exists := orderMap[name]; !exists {
			orderMap[name] = i
		}
	}

	type NameAmount struct {
		Name   string
		Amount int
	}
	var entries []NameAmount

	for name, amount := range merged {
		entries = append(entries, NameAmount{name, amount})
	}

	sort.Slice(entries, func(i, j int) bool {
		if entries[i].Amount == entries[j].Amount {
			return orderMap[entries[i].Name] < orderMap[entries[j].Name]
		}
		return entries[i].Amount > entries[j].Amount
	})

	result := make([]string, len(entries))
	for i, entry := range entries {
		result[i] = entry.Name
	}

	return result
}

func main() {
	fmt.Println(sliceEqual(solution(4, []string{"a", "b", "c", "d"}, []int{1, 2, 2, 1}), []string{"b", "c", "a", "d"}))
	fmt.Println(sliceEqual(solution(3, []string{"x", "y", "z"}, []int{100, 200, 200}), []string{"y", "z", "x"}))
	fmt.Println(sliceEqual(solution(5, []string{"m", "n", "o", "p", "q"}, []int{50, 50, 30, 30, 20}), []string{"m", "n", "o", "p", "q"}))
}

func sliceEqual(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i, v := range a {
		if v != b[i] {
			return false
		}
	}
	return true
}

File: Doubao_MarsCode/java/MC0049/Main.java
package MC0049;

import java.util.*;

public class Main {

    public static List<String> solution(int n, List<String> s, List<Integer> x) {
        Map<String, Integer> merged = new HashMap<>();
        LinkedHashMap<String, Integer> orderMap = new LinkedHashMap<>();

        for (int i = 0; i < n; i++) {
            String name = s.get(i);
            int amount = x.get(i);

            merged.put(name, merged.getOrDefault(name, 0) + amount);

            if (!orderMap.containsKey(name)) {
                orderMap.put(name, i);
            }
        }

        List<Map.Entry<String, Integer>> entries = new ArrayList<>(merged.entrySet());

        entries.sort((a, b) -> {
            int amountA = a.getValue();
            int amountB = b.getValue();

            if (amountA != amountB) {
                return Integer.compare(amountB, amountA);
            }

            return Integer.compare(orderMap.get(a.getKey()), orderMap.get(b.getKey()));
        });

        List<String> result = new ArrayList<>();
        for (Map.Entry<String, Integer> entry : entries) {
            result.add(entry.getKey());
        }

        return result;
    }

    public static void main(String[] args) {
        System.out.println(solution(4, Arrays.asList("a", "b", "c", "d"), Arrays.asList(1, 2, 2, 1)).equals(Arrays.asList("b", "c", "a", "d")));
        System.out.println(solution(3, Arrays.asList("x", "y", "z"), Arrays.asList(100, 200, 200)).equals(Arrays.asList("y", "z", "x")));
        System.out.println(solution(5, Arrays.asList("m", "n", "o", "p", "q"), Arrays.asList(50, 50, 30, 30, 20)).equals(Arrays.asList("m", "n", "o", "p", "q")));
    }
}